common/functions/src/scalars/expressions/cast_with_type.rs

```diff
@@ -166,12 +176,63 @@ pub fn cast_with_type(
     Ok(result)
 }
 
+pub fn cast_to_variant(
+    column: &ColumnRef,
+    from_type: &DataTypePtr,
+) -> Result<(ColumnRef, Option<Bitmap>)> {
+    let column = Series::remove_nullable(column);
+    let size = column.len();
+    let mut builder = ColumnBuilder::<JsonValue>::with_capacity(size);
+
+    with_match_physical_primitive_type!(from_type.data_type_id().to_physical_type(), |$T| {
```

Comment:
StringColumn cast to variant column is not covered.

Reply:
String can't be automatically casted to Variants, it must be explicitly cast using the `to_variant` function